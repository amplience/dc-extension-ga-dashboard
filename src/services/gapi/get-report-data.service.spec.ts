import type { GoogleAnalyticsEmbedAPI } from '../../definitions/google-analytics-embed-api';
import { gapiRequestGuard } from './gapi-request-guard.service';
import { getReportData } from './get-report-data.service';
import { ChartType, insertDataChart } from './insert-data-chart.service';

jest.mock('./gapi-request-guard.service');

describe('getReportData', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should return report data when gapi request guard resolves', async () => {
    (gapiRequestGuard as jest.Mock).mockImplementation((gapi, fn) => fn());
    const onceSuccessMock = jest.fn();
    const onceFailureMock = jest.fn();
    const executeMock = jest.fn();

    const gapi = ({
      analytics: {
        report: {
          Data: jest.fn().mockImplementation(() => ({
            once: onceSuccessMock.mockImplementation((status, resolve) => ({
              once: onceFailureMock.mockImplementation(() => ({
                execute: executeMock.mockImplementation(() => {
                  resolve({
                    rows: [['test-dimension', 1, 2, 3, 4]],
                    totalsForAllResults: {
                      'ga:totalEvents': 3,
                      'ga:uniqueEvents': 2,
                    },
                  });
                }),
              })),
            })),
          })),
        },
      },
    } as unknown) as GoogleAnalyticsEmbedAPI;
    const query = {
      gaViewId: 'ga:test-id',
      dimension: 'ga:test-dimension',
      limit: 100,
      dateRange: { to: '2020-01-01', from: '2020-01-01' },
      gaQueryFilter: undefined,
    };

    const result = await getReportData(gapi, query);

    expect(
      (gapi.analytics.report.Data as jest.Mock).mock.calls
    ).toMatchSnapshot();
    expect(onceSuccessMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "success",
          [Function],
        ],
      ]
    `);
    expect(onceFailureMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "error",
          [Function],
        ],
      ]
    `);
    expect(executeMock).toHaveBeenCalledTimes(1);
    expect(result).toMatchInlineSnapshot(`
      Array [
        Array [
          "test-dimension",
          1,
          33,
          2,
          100,
          3,
          4,
        ],
      ]
    `);
  });

  it('should throw an error if gapi request rejects', async () => {
    (gapiRequestGuard as jest.Mock).mockImplementation((gapi, fn) => fn());
    const onceSuccessMock = jest.fn();
    const onceFailureMock = jest.fn();
    const executeMock = jest.fn();

    const gapi = ({
      analytics: {
        report: {
          Data: jest.fn().mockImplementation(() => ({
            once: onceSuccessMock.mockImplementation(() => ({
              once: onceFailureMock.mockImplementation((status, reject) => ({
                execute: executeMock.mockImplementation(() => {
                  reject(new Error('Gapi request error'));
                }),
              })),
            })),
          })),
        },
      },
    } as unknown) as GoogleAnalyticsEmbedAPI;
    const query = {
      gaViewId: 'ga:test-id',
      dimension: 'ga:test-dimension',
      limit: 100,
      dateRange: { to: '2020-01-01', from: '2020-01-01' },
      gaQueryFilter: undefined,
    };

    const promise = getReportData(gapi, query);

    await expect(() => promise).rejects.toMatchInlineSnapshot(
      `[Error: Gapi request error]`
    );
  });

  it('should throw an error if request guard rejects', async () => {
    const gapi = ({
      analytics: {
        report: {
          Data: jest.fn(),
        },
      },
    } as unknown) as GoogleAnalyticsEmbedAPI;

    (gapiRequestGuard as jest.Mock).mockImplementation(() =>
      Promise.reject(new Error('Gapi auth guard error'))
    );

    const query = {
      gaViewId: 'ga:test-id',
      dimension: 'ga:test-dimension',
      limit: 100,
      dateRange: { to: '2020-01-01', from: '2020-01-01' },
      gaQueryFilter: undefined,
    };

    const promise = getReportData(gapi, query);

    await expect(() => promise).rejects.toMatchInlineSnapshot(
      `[Error: Gapi auth guard error]`
    );
  });
});
