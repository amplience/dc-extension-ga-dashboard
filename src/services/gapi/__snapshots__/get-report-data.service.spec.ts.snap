// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReportData should return report data when gapi request guard resolves 1`] = `
Array [
  Array [
    Object {
      "query": Object {
        "dimensions": "ga:test-dimension",
        "end-date": "2020-01-01",
        "filters": undefined,
        "ids": "ga:test-id",
        "max-results": 100,
        "metrics": "ga:totalEvents,ga:uniqueEvents,ga:eventValue,ga:avgEventValue",
        "sort": "-ga:totalEvents",
        "start-date": "2020-01-01",
      },
    },
  ],
]
`;
